plugins/catalog-backend/src/service/DefaultEntitiesCatalog.ts

```diff
@@ -232,6 +248,119 @@ export class DefaultEntitiesCatalog implements EntitiesCatalog {
     };
   }
 
+  async paginatedEntities(
+    request?: PaginatedEntitiesRequest,
+  ): Promise<PaginatedEntitiesResponse> {
+    const db = this.database;
+    const limit = request?.limit ?? 20;
+
+    const cursor: Omit<Cursor, 'sortFieldId'> & { sortFieldId?: string } = {
+      firstFieldId: '',
+      sortField: defaultSortField,
+      sortFieldOrder: defaultSortFieldOrder,
+      isPrevious: false,
+      ...parseCursorFromRequest(request),
+    };
+
+    const isFetchingBackwards = cursor.isPrevious;
+
+    const dbQuery = db('search')
+      .join('final_entities', 'search.entity_id', 'final_entities.entity_id')
+      .where('key', cursor.sortField);
+
+    if (cursor.filter) {
+      parseFilter(cursor.filter, dbQuery, db, false, 'search.entity_id');
+    }
+
+    const normalizedQueryByName = cursor.query?.trim();
+    if (normalizedQueryByName) {
+      dbQuery.andWhereLike(
+        'value',
+        `%${normalizedQueryByName.toLocaleLowerCase('en-US')}%`,
+      );
+    }
+
+    const countQuery = dbQuery.clone();
+
+    const isOrderingDescending = cursor.sortFieldOrder === 'desc';
+    if (cursor.sortFieldId) {
+      dbQuery.andWhere(
+        'value',
+        isFetchingBackwards !== isOrderingDescending ? '<' : '>',
+        cursor.sortFieldId,
+      );
+    }
```

Comment:
We can't really assume unique values here right? Might need a secondary offset or value to filter by as well? For example always falling back to sorting by the `entity_id`

Reply:
yep, good point. This is related to the comment above of having multiple sortFields

Reply:
Addressed by https://github.com/backstage/backstage/pull/15513